core/irTraversals.ml

```diff
@@ -226,11 +226,21 @@ struct
             let (args, _, o) = o#list (fun o -> o#value) args in
               (* TODO: check arg types match *)
               `ApplyPure (f, args), deconstruct return_type ft, o
-        | `Closure (f, z) ->
+
+        | `Closure (f, tyargs, z) ->
+
             let (f, t, o) = o#var f in
+            let t = if tyargs = []
```

Comment:
Consider using pattern matching rather than if-then-else. I'd probably format the code like this:
```ocaml
let t =
  match tyargs with
  | [] -> t
  | _ ->
     let (remaining_type, instantiation_maps) =
        Instantiate.type_arguments_to_instantiation_maps false t tyargs
     in
     Instantiate.datatype instantiation_maps remaining_type
in
...
```